store/tikv/lock_resolver.go

```diff
@@ -578,10 +581,89 @@ func (lr *LockResolver) getTxnStatus(bo *Backoffer, txnID uint64, primary []byte
 			status.commitTS = cmdResp.CommitVersion
 			lr.saveResolved(txnID, status)
 		}
+
+		if config.GetGlobalConfig().TiKVClient.EnableAsyncCommit && cmdResp.UseAsyncCommit {
+			lr.asyncRecovery()
+		}
+
 		return status, nil
 	}
 }
 
+// resolveLockAsync resolves l assuming it was locked using the async commit protocol. Precondition: status.secondaries is non-empty.
+func (lr *LockResolver) resolveLockAsync(bo *Backoffer, l *Lock, status TxnStatus, cleanRegions map[RegionVerID]struct{}) error {
+	// TODO(nrc) incr a counter?
+
+	regions, _, err := lr.store.GetRegionCache().GroupKeysByRegion(bo, status.secondaries, nil)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	var commitTs uint64 = 0
+	var locks []*Lock
+
+regionLoop:
+	for regionID, keys := range regions {
+		getReq := &kvrpcpb.BatchGetRequest{
+			Keys:    keys,
+			Version: l.TxnID,
+		}
+		req := tikvrpc.NewRequest(tikvrpc.CmdBatchGet, getReq)
+		resp, err := lr.store.SendReq(bo, req, regionID, readTimeoutShort)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		regionErr, err := resp.GetRegionError()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if regionErr != nil {
+			err = bo.Backoff(BoRegionMiss, errors.New(regionErr.String()))
+			if err != nil {
+				return errors.Trace(err)
+			}
+			continue
+		}
+		if resp.Resp == nil {
+			return errors.Trace(ErrBodyMissing)
+		}
+		getResp := resp.Resp.(*kvrpcpb.BatchGetResponse)
+
+		for _, kv := range getResp.Pairs {
+			if e := kv.Error; e != nil && e.Locked != nil {
+				lockInfo := e.Locked
+				if lockInfo.LockVersion == l.TxnID {
+					if lockInfo.MinCommitTs > commitTs {
+						commitTs = lockInfo.MinCommitTs
+					}
+					locks = append(locks, NewLock(kv.Error.Locked))
+					continue
+				}
+			}
+			// kv is not locked or locked by another transaction
+			commitTs = 0
+			break regionLoop
+		}
+	}
+
+	status.commitTS = commitTs
+	// Resolve every lock in the transaction.
+	// TODO can do secondary (but not primary) locks in parallel.
+	for _, lock := range locks {
+		err := lr.resolveLock(bo, lock, status, lite, cleanRegions)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	// Primary lock.
+	err = lr.resolveLock(bo, l, status, lite, cleanRegions)
```

Comment:
Should we resolve the primary key lock first and the secondaries async like the current commit process,so that other transactions could get the transaction status checking only the primary key?

Reply:
It can be a optimization.
If primary fail, the transaction fail.
If primary success, we still need to check all secondary keys to deteminate the transaction state.

Reply:
I think I agree with @tiancaiamao . I thought that we should not have secondary locks which point at a primary lock which has been removed. I think we can optimise later if we are careful